engine/src/main/java/org/pentaho/di/job/Job.java

```diff
@@ -0,0 +1,2367 @@
+      throw new KettleJobException( e ); // In case something else goes wrong.
+    }
+  }
+
+  /**
+   * Writes information to Job Log table. Cleans old records, in case job is finished.
+   */
+  protected void writeLogTableInformation( JobLogTable jobLogTable, LogStatus status ) throws KettleJobException,
+    KettleDatabaseException {
+    boolean cleanLogRecords = status.equals( LogStatus.END );
+    String tableName = jobLogTable.getActualTableName();
+    DatabaseMeta logcon = jobLogTable.getDatabaseMeta();
+
+    Database ldb = createDataBase( logcon );
+    ldb.shareVariablesWith( this );
+    try {
+      ldb.connect();
+      ldb.setCommit( logCommitSize );
+      ldb.writeLogRecord( jobLogTable, status, this, null );
+
+      if ( cleanLogRecords ) {
+        ldb.cleanupLogRecords( jobLogTable, getJobname() );
+      }
+
+    } catch ( KettleDatabaseException dbe ) {
+      addErrors( 1 );
+      throw new KettleJobException( "Unable to end processing by writing log record to table " + tableName, dbe );
+    } finally {
+      if ( !ldb.isAutoCommit() ) {
+        ldb.commitLog( true, jobLogTable );
+      }
+      ldb.disconnect();
+    }
+  }
+
+  /**
+   * Write log channel information.
+   *
+   * @throws KettleException
+   *           the kettle exception
+   */
+  protected void writeLogChannelInformation() throws KettleException {
+    Database db = null;
+    ChannelLogTable channelLogTable = jobMeta.getChannelLogTable();
+
+    // PDI-7070: If parent job has the same channel logging info, don't duplicate log entries
+    Job j = getParentJob();
+
+    if ( j != null ) {
+      if ( channelLogTable.equals( j.getJobMeta().getChannelLogTable() ) ) {
+        return;
+      }
+    }
+    // end PDI-7070
+
+    try {
+      db = new Database( this, channelLogTable.getDatabaseMeta() );
+      db.shareVariablesWith( this );
+      db.connect();
+      db.setCommit( logCommitSize );
+
+      List<LoggingHierarchy> loggingHierarchyList = getLoggingHierarchy();
+      for ( LoggingHierarchy loggingHierarchy : loggingHierarchyList ) {
+        db.writeLogRecord( channelLogTable, LogStatus.START, loggingHierarchy, null );
+      }
+
+      // Also time-out the log records in here...
+      //
+      db.cleanupLogRecords( channelLogTable, getJobname() );
+
+    } catch ( Exception e ) {
+      throw new KettleException( BaseMessages.getString( PKG,
+          "Trans.Exception.UnableToWriteLogChannelInformationToLogTable" ), e );
+    } finally {
+      if ( !db.isAutoCommit() ) {
+        db.commit( true );
+      }
+      db.disconnect();
+    }
+  }
+
+  /**
+   * Write job entry log information.
+   *
+   * @throws KettleException
+   *           the kettle exception
+   */
+  protected void writeJobEntryLogInformation() throws KettleException {
+    Database db = null;
+    JobEntryLogTable jobEntryLogTable = getJobMeta().getJobEntryLogTable();
+    try {
+      db = createDataBase( jobEntryLogTable.getDatabaseMeta() );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *db*. The resource is referenced at lines 1257, 1258, 1259, 1262, 1265, 1270, 1271. The resource is closed at line 1273. There are other execution paths that do not contain closure statements, for example, when Database.commitLog throws an exception, and an if condition at line 1270 evaluates to true.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).